src/test-explorer/test-manager.ts

```diff
@@ -0,0 +1,175 @@
+import * as vscode from "vscode";
+import { TestRunProfileKind, tests } from "vscode";
+import {
+  ExecuteCommandRequest,
+  LanguageClient,
+} from "vscode-languageclient/node";
+import { testCache } from "./test-cache";
+import { runHandler } from "./test-run-handler";
+import {
+  SuiteDiscovery,
+  TargetName,
+  TargetUri,
+  TestDiscoveryResult,
+  TestItemMetadata,
+} from "./types";
+import { toVscodeRange } from "./util";
+
+export function createTestManager(
+  client: LanguageClient,
+  isDisabled: boolean
+): TestManager {
+  return new TestManager(client, isDisabled);
+}
+
+class TestManager {
+  readonly testController = tests.createTestController(
+    "metalsTestController",
+    "Metals Test Explorer"
+  );
+
+  private isDisabled = false;
+  private isRunning = false;
+
+  constructor(private readonly client: LanguageClient, isDisabled: boolean) {
+    if (isDisabled) {
+      this.disable();
+    }
+    this.testController.resolveHandler = async (item?: vscode.TestItem) => {
+      if (item != null) {
+        return;
+      } else {
+        await this.discoverTestSuites();
+      }
+    };
+
+    const callback = () => (this.isRunning = false);
```

Comment:
I experimented a little bit and it seems like this + vscode built-in cancelation mechanism seems to work (more or less). When you are in the middle of test run and you'll click to run another test site `this.isRunning` will prevent you from running the next suite. However, the first request is cancelled and is marked as a skipped/cancelled by vscode even though test-analyzer sets particular test items status. I think it's good enough for start and if this will be a problem we can spend more time on this mechanic. Ok @tgodzik ?